### PR TITLE
Implemented state machine for simulator

### PR DESCRIPTION
**Pull request by Prerna-0202**
#### Describe the changes you have made in this PR -
1. Implemented a state machine with two states: normal and error.
2. The default state is set to "Normal," and the current state is initially assigned as the default state.
3. Implemented a feature to catch uncaught exception errors of the simulator. When an error is caught, the simulator's state is changed from "Normal" to "Error." Additionally, a notification is displayed to the user indicating that the simulator is in an error state.
4.  In the Error state, the user can only delete or undo components. When the erroneous component is removed, the simulator instantly returns to the normal state.

Screenshots:-

Before:-

https://github.com/CircuitVerse/CircuitVerse/assets/89515816/74ce7d87-c8ae-4256-a1fa-a2e9ec411619

After:-

https://github.com/CircuitVerse/CircuitVerse/assets/89515816/8962d660-cab8-41c3-bcfe-66ad519dafea

Note: Please check **Allow edits from maintainers.** If you would like us to assist in the PR.

### CODE REVIEW

File: simulator/src/engine.js [hunk 1]
```diff
@@ -14,11 +14,29 @@ import { showError } from './utils';
...
+    ERROR: 'error',
+};
+
+/**
+ * @type {string}
+ * @category engine
+ */
+const defaultState = states.NORMAL;
+export var currentState = defaultState;
+
```
Comment (pulkit-30) [inline simulator/src/engine.js:39]: This looks good, but is there any reason for not declaring these states under scope class?

Comment (Prerna-0202) [reply simulator/src/engine.js:39]: Nope,  We can also declare this under the scope class. If you want, I can do it.

Comment (Prerna-0202) [reply simulator/src/engine.js]: Done 👍

**Comment by pulkit-30** [comment]
Hey @Prerna-0202 is this PR ready to merge?

**Comment by Prerna-0202** [comment]
> Hey @Prerna-0202 is this PR ready to merge?

No, In this pr, I will work on the suggestions given in the meeting.  Once I am done with [this](https://github.com/CircuitVerse/CircuitVerse/pull/3870) pr.

**Review by pulkit-30 — APPROVED** [review]

**Comment by pulkit-30** [comment]
LGTM 🎉 , Great Work @Prerna-0202

**Pull request merged**